virtinst/virtxml.py

```diff
@@ -63,37 +63,111 @@ def defined_xml_is_unchanged(conn, domain, original_xml):
+
+
+def validate_action(action, conn, options):
+    if options.os_variant is not None:
+        if action.is_edit:
+            fail(_("--os-variant/--osinfo is not supported with --edit"))
+        if action.is_remove_device:
+            fail(_("--os-variant/--osinfo is not supported with --remove-device"))
+        if action.is_build_xml:
+            fail(_("--os-variant/--osinfo is not supported with --build-xml"))
+
+    if not action.parserclass.guest_propname and action.is_build_xml:
+        fail(_("--build-xml not supported for --%s") %
+             action.parserclass.cli_arg_name)
+
+    if action.parserclass is cli.ParserXML and not action.is_edit:
+        fail(_("--xml can only be used with --edit"))
+
+    stub_guest = Guest(conn)
+    if not action.parserclass.prop_is_list(stub_guest):
+        if action.is_remove_device:
+            fail(_("Cannot use --remove-device with --%s") %
+                 action.parserclass.cli_arg_name)
+        if action.is_add_device:
+            fail(_("Cannot use --add-device with --%s") %
+                 action.parserclass.cli_arg_name)
+
+    if options.update and not action.parserclass.guest_propname:
+        fail(_("Don't know how to --update for --%s") %
+             (action.parserclass.cli_arg_name))
 
+
+def check_action_collision(options):
     collisions = []
+    actions = ["edit", "add-device", "remove-device", "build-xml"]
     for cliname in actions:
         optname = cliname.replace("-", "_")
-        if getattr(options, optname) not in [False, -1]:
-            collisions.append(cliname)
+        value = getattr(options, optname)
+        if value not in [False, -1]:
+            collisions.append((cliname, value))
 
     if len(collisions) == 0:
         fail(_("One of %s must be specified.") %
              ", ".join(["--" + c for c in actions]))
     if len(collisions) > 1:
         fail(_("Conflicting options %s") %
-             ", ".join(["--" + c for c in collisions]))
+             ", ".join(["--" + c[0] for c in collisions]))
+
+    action_name, selector = collisions[0]
+    return action_name, selector
 
 
 def check_xmlopt_collision(options):
     collisions = []
     for parserclass in cli.VIRT_PARSERS + [cli.ParserXML]:
-        if getattr(options, parserclass.cli_arg_name):
-            collisions.append(parserclass)
+        value = getattr(options, parserclass.cli_arg_name)
+        if value:
+            collisions.append((parserclass, value))
 
     if len(collisions) == 0:
         fail(_("No change specified."))
     if len(collisions) != 1:
         fail(_("Only one change operation may be specified "
                "(conflicting options %s)") %
-               [c.cli_flag_name() for c in collisions])
+               [c[0].cli_flag_name() for c in collisions])
+
+    parserclass, parservalue = collisions[0]
+    return parserclass, parservalue
 
-    return collisions[0]
+
+def parse_actions(conn, options):
```

Comment:
nitpick: since this function parses the command line options and returns only one Action, I'd name it `parse_action()` (singular); this makes the usage of it more readable:
```python
    action = parse_action(conn, options)
```